components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/SecurePrefsReliabilityExperiment.kt

```diff
@@ -0,0 +1,151 @@
+            } catch (e: Exception) {
+                emitFact(Actions.GET, Values.FAIL, mapOf("javaClass" to e.nameForTelemetry()))
+
+                // should this return? or proceed to the write part..?
+                return
+            }
+
+            val valueAlreadyPersisted = prefs().getBoolean(PREF_DID_STORE_VALUE, false)
+
+            val getResult = when {
+                // we didn't store the value yet, and didn't get anything back either
+                (!valueAlreadyPersisted && storedVal == null) -> {
+                    Values.SUCCESS_MISSING
+                }
+                // we got back the value we stored
+                (valueAlreadyPersisted && storedVal == PREF_VALUE) -> {
+                    Values.SUCCESS_PRESENT
+                }
+                // value was lost
+                (valueAlreadyPersisted && storedVal == null) -> {
+                    Values.LOST
+                }
+                // we got some value back, but not what we stored
+                (valueAlreadyPersisted && storedVal != PREF_VALUE) -> {
+                    Values.CORRUPTED
+                }
+                // we didn't store the value yet, but got something back either way
+                else -> {
+                    Values.PRESENT_UNEXPECTED
+                }
+            }
+
+            emitFact(Actions.GET, getResult)
+
+            when (getResult) {
+                // perform a write of the missing value
+                Values.SUCCESS_MISSING -> {
+                    try {
+                        securePrefs.putString(PREF_KEY, PREF_VALUE)
+                        emitFact(Actions.WRITE, Values.SUCCESS_WRITE)
+                    } catch (e: Exception) {
+                        emitFact(Actions.WRITE, Values.FAIL, mapOf("javaClass" to e.nameForTelemetry()))
+                    }
+                    prefs().edit().putBoolean(PREF_DID_STORE_VALUE, true).apply()
+                }
+                // reset our experiment in case of detected failures. this lets us measure the failure rate
+                Values.LOST, Values.CORRUPTED, Values.PRESENT_UNEXPECTED -> {
+                    securePrefs.clear()
+                    prefs().edit().clear().apply()
+                    emitFact(Actions.RESET, Values.SUCCESS_RESET)
+                }
+                else -> {
+                    // no-op
+                }
+            }
+        } catch (e: Exception) {
+            emitFact(Actions.EXPERIMENT, Values.FAIL, mapOf("javaClass" to e.nameForTelemetry()))
+        }
+    }
+}
+
+private fun emitFact(
+    item: String,
+    value: SecurePrefsReliabilityExperiment.Companion.Values,
+    metadata: Map<String, Any>? = null
+) {
+    Fact(
+        Component.LIB_DATAPROTECT,
+        Action.IMPLEMENTATION_DETAIL,
+        item,
+        "${value.v}",
+        metadata
+    ).collect()
+}
+
+private fun Exception.nameForTelemetry(): String {
+    return this.javaClass.canonicalName ?: "anonymous"
```

Comment:
Is there a size limit if we pass this to Glean? Would be annoying if we only see the beginning of the package name :) Maybe `simpleName` is enough to identify the exception?

Reply:
Hmmmmm their docs says "arbitrary content, except pls no json blobs", so let's take a chance here :)